src/SemiFungible1155.sol

```diff
@@ -0,0 +1,321 @@
+    mapping(uint256 => address) public creators; //TODO extend with admin contracts
+
+    mapping(uint256 => uint256) public tokenValues;
+    mapping(uint256 => uint256) internal maxIndex;
+    mapping(uint256 => mapping(address => uint256)) public tokenUserBalances;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    // solhint-disable-next-line func-name-mixedcase
+    function __SemiFungible1155_init() public virtual initializer {
+        __Upgradeable1155_init();
+    }
+
+    /// ENJIN EXAMPLE IMPLEMENTATION
+    // Only to make code clearer. Should not be functions
+    // TODO cleanup functions
+    function isNonFungible(uint256 _id) internal pure returns (bool) {
+        return _id & TYPE_NF_BIT == TYPE_NF_BIT;
+    }
+
+    function isFungible(uint256 _id) internal pure returns (bool) {
+        return _id & TYPE_NF_BIT == 0;
+    }
+
+    function getNonFungibleIndex(uint256 _id) internal pure returns (uint256) {
+        return _id & NF_INDEX_MASK;
+    }
+
+    function getNonFungibleBaseType(uint256 _id) internal pure returns (uint256) {
+        return _id & TYPE_MASK;
+    }
+
+    function isNonFungibleBaseType(uint256 _id) internal pure returns (bool) {
+        // A base type has the NF bit but does not have an index.
+        return (_id & TYPE_NF_BIT == TYPE_NF_BIT) && (_id & NF_INDEX_MASK == 0);
+    }
+
+    function isNonFungibleItem(uint256 _id) internal pure returns (bool) {
+        // A base type has the NF bit but does has an index.
+        return (_id & TYPE_NF_BIT == TYPE_NF_BIT) && (_id & NF_INDEX_MASK != 0);
+    }
+
+    /// READ
+
+    function balanceOf(address _owner, uint256 _tokenID) public view override returns (uint256 tokenUserBalance) {
+        tokenUserBalance = tokenUserBalances[_tokenID][_owner];
+    }
+
+    function balanceOf(uint256 _tokenID) public view returns (uint256 tokenValue) {
+        tokenValue = tokenValues[_tokenID];
+    }
+
+    function totalSupply(uint256 _typeID) external view returns (uint256 total) {
+        total = tokenValues[_typeID];
+    }
+
+    /// MUTATE
+
+    /// @dev create token type ID based of token counter
+    // TODO should creator be msg.sender or inital account?
+    function _createTokenType() internal returns (uint256 typeID) {
+        typeID = (++typeCounter << 128); //TODO max value check?
+        creators[typeID] = msg.sender;
+    }
+
+    /// @dev create token type ID based of token counter
+    function _createTokenType(uint256 units, string memory uri) internal returns (uint256 typeID) {
+        typeID = (++typeCounter << 128); //TODO max value check?
+        _setURI(typeID, uri);
+
+        creators[typeID] = msg.sender;
+        tokenValues[typeID] = units;
+    }
+
+    /// @dev Mint a new token type and the initial value
+    function _mintValue(address _account, uint256 _value, string memory uri) internal returns (uint256 typeID) {
```

Comment:
it feels a little bit off to have `uri` a mandatory parameter here (even though it's practical for your usecase). From a code perspective, I as a consumer of this primitive would decide on my own whether I'd need token uris or e.g. generate metadata on chain (as you did before)